File: app/src/main/java/org/mifos/mobilebanking/ui/adapters/LoanAccountsListAdapter.java
package org.mifos.mobilebanking.ui.adapters;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.injection.ActivityContext;
import org.mifos.mobilebanking.models.accounts.loan.LoanAccount;
import org.mifos.mobilebanking.utils.CurrencyUtil;
import org.mifos.mobilebanking.utils.DateHelper;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Vishwajeet
 * @since 22/6/16.
 */
public class LoanAccountsListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final Context context;
    private List<LoanAccount> loanAccountsList = new ArrayList<>();

    @Inject
    public LoanAccountsListAdapter(@ActivityContext Context context) {
        this.context = context;
    }

    public void setLoanAccountsList(List<LoanAccount> loanAccountsList) {
        this.loanAccountsList = loanAccountsList;
        notifyDataSetChanged();
    }

    public List<LoanAccount> getLoanAccountsList() {
        return loanAccountsList;
    }

    public LoanAccount getItem(int position) {
        return loanAccountsList.get(position);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder vh;
        View v = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.row_loan_account, parent, false);
        vh = new ViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ViewHolder) {

            LoanAccount loanAccount = getItem(position);
            ((ViewHolder) holder).tvClientLoanAccountNumber.setText(loanAccount.getAccountNo());
            ((ViewHolder) holder).tvLoanAccountProductName.setText(loanAccount.getProductName());
            ((ViewHolder) holder).tvAccountBalance.setVisibility(View.GONE);

            if (loanAccount.getStatus().getActive() && loanAccount.getInArrears()) {

                setLoanAccountsGeneralDetails(holder, R.color.red, context.getString(R.string.
                        string_and_string, context.getString(R.string.disbursement), DateHelper
                        .getDateAsString(loanAccount.getTimeline().getActualDisbursementDate())));
                setLoanAccountsDetails(((ViewHolder) holder), loanAccount, R.color.red);

            } else if (loanAccount.getStatus().getActive()) {

                setLoanAccountsGeneralDetails(holder, R.color.deposit_green, context.getString(R.
                        string.string_and_string, context.getString(R.string.disbursement),
                        DateHelper.getDateAsString(loanAccount.getTimeline().
                                getActualDisbursementDate())));
                setLoanAccountsDetails(((ViewHolder) holder), loanAccount, R.color.deposit_green);

            } else if (loanAccount.getStatus().getWaitingForDisbursal()) {

                setLoanAccountsGeneralDetails(holder, R.color.blue, context.getString(R.string.
                        string_and_string, context.getString(R.string.approved), DateHelper
                        .getDateAsString(loanAccount.getTimeline().getApprovedOnDate())));

            } else if (loanAccount.getStatus().getPendingApproval()) {

                setLoanAccountsGeneralDetails(holder, R.color.light_yellow, context.getString(R.
                                string.string_and_string, context.getString(R.string.
                                submitted), DateHelper.getDateAsString(loanAccount.
                                getTimeline().getSubmittedOnDate())));

            }  else if (loanAccount.getStatus().getOverpaid()) {

                setLoanAccountsDetails(((ViewHolder) holder), loanAccount, R.color.purple);
                setLoanAccountsGeneralDetails(holder, R.color.purple, context.getString(R.string.
                        string_and_string, context.getString(R.string.disbursement), DateHelper
                        .getDateAsString(loanAccount.getTimeline().getActualDisbursementDate())));

            } else if (loanAccount.getStatus().getClosed()) {

                setLoanAccountsGeneralDetails(holder, R.color.black, context.getString(R.string.
                        string_and_string, context.getString(R.string.closed), DateHelper
                        .getDateAsString(loanAccount.getTimeline().getClosedOnDate())));

            } else {
                setLoanAccountsGeneralDetails(holder, R.color.gray_dark, context.getString(R.string.
                            string_and_string, context.getString(R.string.withdrawn), DateHelper
                        .getDateAsString(loanAccount.getTimeline().getWithdrawnOnDate())));

            }
        }

    }

    private void setLoanAccountsDetails(ViewHolder viewHolder, LoanAccount loanAccount, int color) {

        double amountBalance = loanAccount.getLoanBalance() != 0 ? loanAccount.getLoanBalance() : 0;
        viewHolder.tvAccountBalance.setVisibility(View.VISIBLE);
        viewHolder.tvAccountBalance.setText(CurrencyUtil.formatCurrency(context, amountBalance));
        viewHolder.tvAccountBalance.setTextColor(ContextCompat.getColor(context, color));
    }

    private void setLoanAccountsGeneralDetails(RecyclerView.ViewHolder holder, int colorId,
                                                 String dateStr) {
        ((ViewHolder) holder).ivStatusIndicator.setColorFilter(ContextCompat.
                getColor(context, colorId));
        ((ViewHolder) holder).tvDate.setText(dateStr);
    }
    @Override
    public int getItemCount() {
        return loanAccountsList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.tv_clientLoanAccountNumber)
        TextView tvClientLoanAccountNumber;

        @BindView(R.id.tv_loanAccountProductName)
        TextView tvLoanAccountProductName;

        @BindView(R.id.iv_status_indicator)
        AppCompatImageView ivStatusIndicator;

        @BindView(R.id.tv_date)
        TextView tvDate;

        @BindView(R.id.tv_account_balance)
        TextView tvAccountBalance;

        public ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }
    }
}

File: app/src/main/java/org/mifos/mobile/ui/adapters/LoanAccountsListAdapter.kt
package org.mifos.mobile.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import org.mifos.mobile.R
import org.mifos.mobile.core.model.entity.accounts.loan.LoanAccount
import org.mifos.mobile.databinding.RowLoanAccountBinding
import org.mifos.mobile.core.common.utils.CurrencyUtil.formatCurrency
import org.mifos.mobile.core.common.utils.DateHelper.getDateAsString

/**
 * @author Vishwajeet
 * @since 22/6/16.
 */
class LoanAccountsListAdapter(
    val onItemClick: (itemPosition: Int) -> Unit,
) : RecyclerView.Adapter<LoanAccountsListAdapter.ViewHolder>() {

    private var loanAccountsList: List<LoanAccount?>? = ArrayList()

    fun setLoanAccountsList(loanAccountsList: List<LoanAccount?>?) {
        this.loanAccountsList = loanAccountsList
        notifyDataSetChanged()
    }

    fun getLoanAccountsList(): List<LoanAccount?>? {
        return loanAccountsList
    }

    fun getItem(position: Int): LoanAccount? {
        return loanAccountsList?.get(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            RowLoanAccountBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val loanAccount = getItem(position)
        holder.bind(loanAccount)
    }

    override fun getItemCount(): Int {
        return if (loanAccountsList != null) {
            loanAccountsList!!.size
        } else {
            0
        }
    }

    inner class ViewHolder(private val binding: RowLoanAccountBinding) :
        RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.setOnClickListener {
                onItemClick(bindingAdapterPosition)
            }
        }

        fun bind(loanAccount: LoanAccount?) {
            with(binding) {
                tvClientLoanAccountNumber.text = loanAccount?.accountNo
                tvLoanAccountProductName.text = loanAccount?.productName
                tvAccountBalance.visibility = View.GONE
                val context = itemView.context
                if (loanAccount?.status?.active == true && loanAccount.inArrears == true) {
                    setLoanAccountsGeneralDetails(
                        R.color.red,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.disbursement),
                            getDateAsString(loanAccount.timeline?.actualDisbursementDate),
                        ),
                    )
                    setLoanAccountsDetails(loanAccount, R.color.red)
                } else if (loanAccount?.status?.active == true) {
                    setLoanAccountsGeneralDetails(
                        R.color.deposit_green,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.disbursement),
                            getDateAsString(loanAccount.timeline?.actualDisbursementDate),
                        ),
                    )
                    setLoanAccountsDetails(loanAccount, R.color.deposit_green)
                } else if (loanAccount?.status?.waitingForDisbursal == true) {
                    setLoanAccountsGeneralDetails(
                        R.color.blue,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.approved),
                            getDateAsString(loanAccount.timeline?.approvedOnDate),
                        ),
                    )
                } else if (loanAccount?.status?.pendingApproval == true) {
                    setLoanAccountsGeneralDetails(
                        R.color.light_yellow,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.submitted),
                            getDateAsString(loanAccount.timeline?.submittedOnDate),
                        ),
                    )
                } else if (loanAccount?.status?.overpaid == true) {
                    setLoanAccountsDetails(loanAccount, R.color.purple)
                    setLoanAccountsGeneralDetails(
                        R.color.purple,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.disbursement),
                            getDateAsString(loanAccount.timeline?.actualDisbursementDate),
                        ),
                    )
                } else if (loanAccount?.status?.closed == true) {
                    setLoanAccountsGeneralDetails(
                        R.color.black,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.closed),
                            getDateAsString(loanAccount.timeline?.closedOnDate),
                        ),
                    )
                } else {
                    setLoanAccountsGeneralDetails(
                        R.color.gray_dark,
                        context.getString(
                            R.string.string_and_string,
                            context.getString(R.string.withdrawn),
                            getDateAsString(loanAccount?.timeline?.withdrawnOnDate),
                        ),
                    )
                }
            }
        }

        private fun setLoanAccountsDetails(loanAccount: LoanAccount, color: Int) {
            with(binding) {
                val amountBalance: Double =
                    if (loanAccount.loanBalance != 0.0) loanAccount.loanBalance else 0.0
                tvAccountBalance.visibility = View.VISIBLE
                tvAccountBalance.text = formatCurrency(itemView.context, amountBalance)
                tvAccountBalance.setTextColor(ContextCompat.getColor(itemView.context, color))
            }
        }

        private fun setLoanAccountsGeneralDetails(colorId: Int, dateStr: String) {
            with(binding) {
                ivStatusIndicator.setColorFilter(ContextCompat.getColor(itemView.context, colorId))
                tvDate.text = dateStr
            }
        }
    }
}
